File: scripts/download-audio.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
音频文件下载脚本 - 备用方案
使用Python下载免费音频资源

音频来源：
- Pixabay Music: https://pixabay.com/music/
- FreePD: https://freepd.com/
- Freesound: https://freesound.org/
"""

import os
import sys
import urllib.request
import urllib.error
from pathlib import Path

# 设置控制台编码
if sys.platform == "win32":
    import codecs
    sys.stdout = codecs.getwriter("utf-8")(sys.stdout.buffer, "strict")
    sys.stderr = codecs.getwriter("utf-8")(sys.stderr.buffer, "strict")

# 项目根目录
PROJECT_ROOT = Path(__file__).parent.parent
AUDIO_DIR = PROJECT_ROOT / "public" / "assets" / "audio"

# 音频文件配置
AUDIO_FILES = {
    "bgm": [
        {
            "name": "tension.mp3",
            "url": "https://cdn.pixabay.com/download/audio/2022/03/10/audio_5b357619d0.mp3",
            "description": "Suspense tension music"
        },
        {
            "name": "conspiracy.mp3",
            "url": "https://cdn.pixabay.com/download/audio/2022/01/18/audio_d0a13f69d0.mp3",
            "description": "Mysterious atmosphere music"
        },
        {
            "name": "dramatic.mp3",
            "url": "https://cdn.pixabay.com/download/audio/2022/03/24/audio_21fc663520.mp3",
            "description": "Dramatic cinematic music"
        },
    ],
    "sfx": [
        {
            "name": "coin-clink.mp3",
            "url": "https://cdn.pixabay.com/download/audio/2021/08/04/audio_49667085b0.mp3",
            "description": "Coin sound effect"
        },
        {
            "name": "bell.mp3",
            "url": "https://cdn.pixabay.com/download/audio/2022/03/10/audio_13a4412832.mp3",
            "description": "Bell sound"
        },
        {
            "name": "impact.mp3",
            "url": "https://cdn.pixabay.com/download/audio/2021/08/08/audio_06b6288505.mp3",
            "description": "Impact sound effect"
        },
    ]
}

def download_file(url: str, dest_path: Path) -> bool:
    """下载文件到指定路径"""
    try:
        print(f"  [DOWNLOAD] {url}")
        
        # 创建请求
        req = urllib.request.Request(
            url,
            headers={
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        )
        
        # 下载文件
        with urllib.request.urlopen(req, timeout=30) as response:
            with open(dest_path, 'wb') as f:
                f.write(response.read())
        
        print(f"  [OK] Saved: {dest_path}")
        return True
        
    except urllib.error.HTTPError as e:
        print(f"  [ERROR] HTTP {e.code}: {e.reason}")
        return False
    except urllib.error.URLError as e:
        print(f"  [ERROR] URL: {e.reason}")
        return False
    except Exception as e:
        print(f"  [ERROR] Failed: {e}")
        return False

def main():
    print("Music File Downloader for Currency War Animation\n")
    
    # 创建目录
    for category in ["bgm", "sfx", "voiceover"]:
        category_dir = AUDIO_DIR / category
        category_dir.mkdir(parents=True, exist_ok=True)
        print(f"[DIR] Created: {category_dir}")
    
    print("\n[DOWNLOAD] Starting...\n")
    
    success_count = 0
    fail_count = 0
    
    # 下载每个分类的文件
    for category, files in AUDIO_FILES.items():
        print(f"\n--- {category.upper()} ---")
        
        for audio in files:
            dest_path = AUDIO_DIR / category / audio["name"]
            
            if download_file(audio["url"], dest_path):
                print(f"   [INFO] {audio['description']}\n")
                success_count += 1
            else:
                print(f"   [SKIP] {audio['name']}\n")
                fail_count += 1
                
                # 创建占位文件
                dest_path.write_text("# Placeholder - download failed\n")
    
    print("\n" + "=" * 50)
    print(f"[SUCCESS] Downloaded: {success_count} files")
    print(f"[FAILED] {fail_count} files")
    print("=" * 50)
    
    if fail_count > 0:
        print("\n[TIPS] Some files failed to download. You can:")
        print("   1. Manually download from:")
        print("      - Pixabay Music: https://pixabay.com/music/")
        print("      - FreePD: https://freepd.com/")
        print(f"   2. Place in: {AUDIO_DIR}")
    
    print("\n[DONE] Complete!")

if __name__ == "__main__":
    main()

File: scripts/download-audio.ts
#!/usr/bin/env node
/**
 * 自动下载音频文件
 *
 * 从免费音频库下载背景音乐和音效
 * 音频来源：
 * - FreePD: https://freepd.com/
 * - Incompetech: https://incompetech.com/
 * - YouTube Audio Library (无版权音乐)
 */

import { writeFileSync, mkdirSync } from "fs";
import { join } from "path";
import https from "https";
import http from "http";

const AUDIO_DIR = join(process.cwd(), "public", "assets", "audio");

// 音频文件列表
const AUDIO_FILES = [
  // 背景音乐 - 悬疑紧张类
  {
    category: "bgm",
    name: "tension.mp3",
    url: "https://incompetech.com/music/royalty-free/mp3-royaltyfree/Tense.mp3",
    description: "Suspense tension music for dramatic scenes",
  },
  {
    category: "bgm",
    name: "conspiracy.mp3",
    url: "https://incompetech.com/music/royalty-free/mp3-royaltyfree/Disorienting.mp3",
    description: "Mysterious conspiracy atmosphere music",
  },
  {
    category: "bgm",
    name: "epic-main.mp3",
    url: "https://incompetech.com/music/royalty-free/mp3-royaltyfree/BossBattle.mp3",
    description: "Epic battle theme for Waterloo scene",
  },

  // 音效 - 战争场景
  {
    category: "sfx",
    name: "cannon_fire.mp3",
    url: "https://www.freesound.org/data/previews/320/320569_5270888-lq.mp3",
    description: "Cannon fire sound effect",
  },
  {
    category: "sfx",
    name: "war_drums.mp3",
    url: "https://www.freesound.org/data/previews/393/393708_7291208-lq.mp3",
    description: "War drums sound",
  },

  // 音效 - 交易所场景
  {
    category: "sfx",
    name: "stock_bell.mp3",
    url: "https://www.freesound.org/data/previews/341/341695_5858076-lq.mp3",
    description: "Stock exchange bell",
  },
  {
    category: "sfx",
    name: "coin-clink.mp3",
    url: "https://www.freesound.org/data/previews/145/145945_2398804-lq.mp3",
    description: "Gold coin clinking sound",
  },

  // 环境音效
  {
    category: "sfx",
    name: "crowd_murmur.mp3",
    url: "https://www.freesound.org/data/previews/534/534632_12513884-lq.mp3",
    description: "Crowd murmuring and panic",
  },
];

/**
 * 下载文件
 */
function downloadFile(url: string, destPath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith("https") ? https : http;

    console.log(`  ⬇️  Downloading: ${url}`);

    protocol
      .get(url, (response) => {
        // 处理重定向
        if (response.statusCode === 301 || response.statusCode === 302) {
          const redirectUrl = response.headers.location;
          if (redirectUrl) {
            console.log(`  ↪️  Redirecting to: ${redirectUrl}`);
            downloadFile(redirectUrl, destPath).then(resolve).catch(reject);
            return;
          }
        }

        // 检查响应状态
        if (response.statusCode !== 200) {
          reject(new Error(`Failed to download: ${response.statusCode}`));
          return;
        }

        const data: Buffer[] = [];

        response.on("data", (chunk) => {
          data.push(chunk);
        });

        response.on("end", () => {
          const buffer = Buffer.concat(data);
          writeFileSync(destPath, buffer);
          console.log(`  ✅ Saved: ${destPath}`);
          resolve();
        });
      })
      .on("error", (err) => {
        reject(err);
      });
  });
}

/**
 * 主函数
 */
async function main() {
  console.log("🎵 Audio File Downloader for Currency War Animation\n");

  // 创建目录
  const categories = ["bgm", "sfx", "voiceover"];
  categories.forEach((cat) => {
    const dir = join(AUDIO_DIR, cat);
    try {
      mkdirSync(dir, { recursive: true });
      console.log(`📁 Created directory: ${dir}`);
    } catch (err: any) {
      if (err.code !== "EEXIST") {
        console.error(`❌ Failed to create directory: ${dir}`);
      }
    }
  });

  console.log("\n📥 Starting downloads...\n");

  let successCount = 0;
  let failCount = 0;

  // 下载每个文件
  for (const audio of AUDIO_FILES) {
    const destPath = join(AUDIO_DIR, audio.category, audio.name);

    try {
      await downloadFile(audio.url, destPath);
      console.log(`   📝 ${audio.description}\n`);
      successCount++;
    } catch (err: any) {
      console.error(`  ❌ Failed to download ${audio.name}: ${err.message}\n`);
      failCount++;

      // 创建一个占位文件，避免引用错误
      try {
        writeFileSync(destPath, "// Placeholder - download failed\n");
        console.log(`  💾 Created placeholder file\n`);
      } catch (e) {
        // Ignore
      }
    }
  }

  console.log("\n" + "=".repeat(50));
  console.log(`✅ Successfully downloaded: ${successCount} files`);
  console.log(`❌ Failed: ${failCount} files`);
  console.log("=".repeat(50));

  if (failCount > 0) {
    console.log("\n⚠️  Some files failed to download.");
    console.log("💡 Alternative: Manually download from:");
    console.log("   - FreePD: https://freepd.com/");
    console.log("   - Incompetech: https://incompetech.com/");
    console.log("   - Freesound: https://freesound.org/");
    console.log("\n   Then place them in: public/assets/audio/{category}/");
  }

  console.log("\n🎉 Done!");
}

// 运行主函数
main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});
